File: solution/c#/Day24/Day24.Tests/SubmarineTests.cs
using FluentAssertions;
using Xunit;
using static System.IO.File;

namespace Day24.Tests
{
    // Original specifications available here : https://adventofcode.com/2021/day/2
    public class SubmarineTests
    {
        [Fact]
        public void Should_Move_On_Given_Instructions()
            => new K991_P()
                .ToString(LoadInstructions())
                .CalculateResult()
                .Should()
                .Be(1_690_020);

        private static IEnumerable<A> LoadInstructions()
            => ReadAllLines("submarine.txt")
                .Select(A.f);
    }

    internal static class SubmarineExtensions
    {
        public static int CalculateResult(this K991_P k991P)
            => k991P.B().b * k991P.B().a;
    }
}

File: exercise/kotlin/day24/src/test/kotlin/SubmarineTests.kt
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import submarine.Instruction
import submarine.Position
import submarine.Submarine
import submarine.toInstruction
import java.io.File

// Original specifications available here : https://adventofcode.com/2021/day/2
class SubmarineTests : StringSpec({
    "should move on given instructions" {
        loadInstructions()
            .let {
                Submarine(Position())
                    .move(it)
                    .calculateResult() shouldBe 1_690_020
            }
    }
})

private fun Submarine.calculateResult(): Int = position.depth * position.horizontal

private fun loadInstructions(): List<Instruction> =
    File(ClassLoader.getSystemResource("submarine.txt").file)
        .readLines()
        .map { it.toInstruction() }